### PR TITLE
[Tablet Orders] Use recalculate button for selection in side-by-side mode

### LINKED ISSUE #11908: [Tablet Orders] Improve product selection/deselection behaviour to reduce sync errors
Potential approaches:

- Block product selection during "major updates" (current approach, inheirited from iPhone)
- Block product selection for every update (should be reliable, but quite frustrating to use. Probably our quickest option for M3)
- Cancel or ignore API requests that update the order whenever we send a new request (explored in dc044c1)
- Keep a copy of the latest in-flight request and use that as the basis of the order summary and selection changes (started exploring in eb3033e)
- Implement a request queue and ensure they all get sent/applied
- Refactor/wrap RemoteOrderSynchronizer to make it the source of truth for the view models (editable order, product selector, variation selector)
- Have another shared source of truth for the view models – essentially one class which accepts change requests from any of the three views, and maps them to an order request. Vends the selected items for correct display on the three views. May combine with some of the above approaches. Would be good to pull some of the Product/ItemID-based logic out of the view models.

See p1707147517909389-slack-C025A8VV728 for more context

### PR DESCRIPTION
**Pull request by joshheald**
Part of: #11908

## Description

This PR moves our side-by-side order creation experience to use a `Recalculate` button, instead of the `immediate` sync approach. Users must tap `Recalculate` in side-by-side mode to sync the order changes before they can continue with creating the order.

It’s added as temporary mitigation for selection sync difficulties.

We’ll go back to immediate in the near future, so that code is left in for ease later.

Note that this PR also moves the responsibility for triggering sync to the EditableOrderViewModel, via the previously existing closures such as `onProductSelectionStateChanged` or `onAllSelectionsCleared`. This seems like less of a side-effect than having the Product Selector call `onMultipleSelectionCompleted` at appropriate times.

## Testing instructions

1. Enable the `sideBySideViewForOrderForm` feature flag
2. Launch the app on an iPad or large iPhone in landscape
3. Navigate to the `Orders` tab, and tap `+` to open the order form
4. Add products to the order
5. Observe that the button shows `Recalculate`
6. Tap `Recalculate` – observe that the order summary is updated, and the completion button becomes `Collect Payment`.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/2472348/df7e3c2d-9a78-4844-a598-f18575edec83

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/188">17.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift [hunk 1]
```diff
@@ -289,6 +289,75 @@ final class EditableOrderViewModelTests: XCTestCase {
...
+        viewModel.selectionSyncApproach = .onRecalculateButtonTap
+        let productSelectorViewModel = try XCTUnwrap(viewModel.productSelectorViewModel)
+
+        // When
+        productSelectorViewModel.changeSelectionStateForProduct(with: product.productID)
+
+        // Then
+        switch viewModel.doneButtonType {
+        case .recalculate:
+            // Success – we just don't care about the `loading` parameter
```
Comment (iamgabrielma) [inline WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift:330]: TIL we don't need to pass any associated value for testing, and we can just pattern-match it, then make the test succeed when we don't land on the XCTFail case.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift [hunk 2]
```diff
@@ -181,7 +181,7 @@ struct OrderForm: View {
     var body: some View {
         orderFormSummary(presentProductSelector)
             .onAppear {
-                viewModel.syncChangesImmediately = presentationStyle == .sideBySide
+                viewModel.selectionSyncApproach = presentationStyle == .sideBySide ? .onRecalculateButtonTap : .onSelectorButtonTap
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift]: Would it be more readable if we move this bit to a computed variable outside of the view's body?
```    
private var selectionSyncApproach: EditableOrderViewModel.OrderItemSelectionSyncApproach {
        presentationStyle == .sideBySide ? .onRecalculateButtonTap : .onSelectorButtonTap
    }

    var body: some View {
        orderFormSummary(presentProductSelector)
            .onAppear {
                viewModel.selectionSyncApproach = selectionSyncApproach
            }
    }
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 3]
```diff
@@ -1227,24 +1250,63 @@ private extension EditableOrderViewModel {
...
+                case .editing:
                     return .none
                 }
             }
             .assign(to: &$navigationTrailingItem)
     }
 
+    /// Calculates what Call to Action button should be shown depending on our internal state.
+    ///
+    func configureDoneButton() {
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:1284]: Really nice way to setup both the done button and navigation items  💯 I need to copy this approach in those places where the rendering depends on multiple computed variables.

**Review by iamgabrielma — APPROVED** [review]
Works great! :shipit: 

At some point during testing I managed to break the variations selected state somehow (none selected, but on tapping recalculate added them to the order) but I haven't been able to reproduce it. 

Definitely an improvement UX-wise, as well as moving the sync responsibility up the chain from the selector view to the order creation 💯

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift:330]: I still don't love it; the implicit pass isn't that nice to read. It's good to ignore the associated value though (without writing an `==` implementation that ignores it – I ran across that in a test the other day, we'd added it in the past and so my test which _did_ care about the associated value was passing even when it shouldn't have!)

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift]: I'll take a look, but probably still on the view somewhere. It's about converting `presentationStyle` to `syncApproach`, and it feels like it fits better in the OrderForm view, which covers both the relevant views.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 3 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:1284]: I just updated and copied it, tbh, but it is a nice contained use of combine which isn't _too_ convoluted. Breaking it down into separate publishers for the `isPerformingNetworkRequest` concept helped.

**Comment by joshheald** [comment]
> Works great! :shipit:

Thanks for the review!

> At some point during testing I managed to break the variations selected state somehow (none selected, but on tapping recalculate added them to the order) but I haven't been able to reproduce it.

Yep, this is totally possible. I meant to mention it in the description but forgot, sorry!

It predates this PR, so I added a separate ticket for it: #12135

**Pull request merged**